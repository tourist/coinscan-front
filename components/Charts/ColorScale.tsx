import { useEffect } from 'react';
import Box from '@mui/material/Box';
import { select as d3_select } from 'd3';

function renderColorScale(
  selector: string,
  data: number | number[],
  userOptions?: {}
) {
  let negativePercentText: number | undefined;
  let positivePercentText: number | undefined;
  let negativePercent: number | undefined;
  let positivePercent: number | undefined;

  const defaultOptions = {
    textMargin: 2,
    textMarginSwitchThreshold: 45,
  };

  const options = {
    ...defaultOptions,
    ...userOptions,
  };

  if (Array.isArray(data)) {
    if (data.length === 2) {
      negativePercentText = data[0];
      negativePercent = Math.max(data[0], -100);
      positivePercentText = data[1];
      positivePercent = Math.min(data[1], 100);
    } else {
      console.error(
        `Wrong data array length. Expected 2 elements. Passed: ${data.length}`
      );
      return;
    }
  } else {
    if (data === 0) {
      console.warn(
        "0 passed as data value, should be handled outside as special case as it's not presentable."
      );
      return;
    } else if (data < 0) {
      negativePercentText = data;
      negativePercent = Math.max(-100, data);
    } else {
      positivePercentText = data;
      positivePercent = Math.min(data, 100);
    }
  }

  let negativeTextMargin = options.textMargin;
  let positiveTextMargin = -options.textMargin;
  let negativeTextAnchor = 'start';
  let positiveTextAnchor = 'end';

  if (negativePercent && negativePercent > -options.textMarginSwitchThreshold) {
    negativeTextAnchor = 'end';
    negativeTextMargin = -negativeTextMargin;
  }

  if (positivePercent && positivePercent < options.textMarginSwitchThreshold) {
    positiveTextAnchor = 'start';
    positiveTextMargin = options.textMargin;
  }

  const negativePercentDisplay =
    negativePercent && Math.abs(negativePercent / 2);
  const positivePercentDisplay = positivePercent && positivePercent / 2;

  // clear existing renders
  d3_select(selector).selectAll('svg').remove();

  const svg = d3_select(selector)
    .append('svg')
    .attr('viewBox', '0 0 150 20')
    .attr('background', 'black')
    .attr('preserveAspectRatio', 'xMidYMid');

  const defs = svg.append('defs');

  const sanitizedSelector = selector.replace('#', '');

  if (negativePercentDisplay) {
    defs
      .append('clipPath')
      .attr('id', `negative-clip-${sanitizedSelector}`)
      .append('rect')
      .attr('x', `${50 - negativePercentDisplay}%`)
      .attr('y', 0)
      .attr('width', `${negativePercentDisplay}%`)
      .attr('height', '100%');
    svg
      .append('rect')
      .attr('x', 0)
      .attr('y', 0)
      .attr('width', '50%')
      .attr('height', '100%')
      .attr('fill', `url(#negative-gradient-${sanitizedSelector})`)
      .attr('clip-path', `url(#negative-clip-${sanitizedSelector})`);

    const gradientNegative = defs
      .append('linearGradient')
      .attr('id', 'negative-gradient')
      .attr('x1', '0%')
      .attr('x2', '100%')
      .attr('y1', '0%')
      .attr('y2', '0%')
      .attr('id', `negative-gradient-${sanitizedSelector}`);

    gradientNegative
      .append('stop')
      .attr('class', 'start')
      .attr('offset', '0%')
      .attr('stop-color', 'red')
      .attr('stop-opacity', 1);

    gradientNegative
      .append('stop')
      .attr('class', 'end')
      .attr('offset', '100%')
      .attr('stop-color', '#ff7377')
      .attr('stop-opacity', 1);
    svg
      .append('text')
      .text(
        `${
          negativePercentText ? parseFloat(negativePercentText.toFixed(2)) : '-'
        }%`
      )
      .attr('dominant-baseline', 'central')
      .attr('text-anchor', negativeTextAnchor)
      .attr('x', `${50 - negativePercentDisplay + negativeTextMargin}%`)
      .attr('y', '50%')
      .attr('font-size', '0.7rem')
      .attr('font-weight', '700')
      .attr('fill', '#ffffff');
  }

  if (positivePercentDisplay) {
    defs
      .append('clipPath')
      .attr('id', `positive-clip-${sanitizedSelector}`)
      .append('rect')
      .attr('x', `50%`)
      .attr('y', 0)
      .attr('width', `${positivePercentDisplay}%`)
      .attr('height', '100%');

    svg
      .append('rect')
      .attr('x', '50%')
      .attr('y', 0)
      .attr('width', '50%')
      .attr('height', '100%')
      .attr('fill', `url(#positive-gradient-${sanitizedSelector})`)
      .attr('clip-path', `url(#positive-clip-${sanitizedSelector})`);

    const gradientPositive = defs
      .append('linearGradient')
      .attr('x1', '0%')
      .attr('x2', '100%')
      .attr('y1', '0%')
      .attr('y2', '0%')
      .attr('id', `positive-gradient-${sanitizedSelector}`);
    gradientPositive
      .append('stop')
      .attr('class', 'start')
      .attr('offset', '0%')
      .attr('stop-color', '#77dd77')
      .attr('stop-opacity', 1);
    gradientPositive
      .append('stop')
      .attr('class', 'end')
      .attr('offset', '100%')
      .attr('stop-color', '#09b051')
      .attr('stop-opacity', 1);

    svg
      .append('text')
      .text(
        `+${
          positivePercentText ? parseFloat(positivePercentText.toFixed(2)) : '-'
        }%`
      )
      .attr('dominant-baseline', 'central')
      .attr('text-anchor', positiveTextAnchor)
      .attr('x', `${50 + positivePercentDisplay + positiveTextMargin}%`)
      .attr('y', '50%')
      .attr('font-size', '0.7rem')
      .attr('font-weight', '700')
      .attr('fill', '#ffffff');
  }
}

type ColorScaleProps = {
  id: number | string;
  data: number | number[];
};

const ColorScale = ({ id, data }: ColorScaleProps) => {
  useEffect(() => renderColorScale(`#color-box${id}`, data), [id, data]);
  return (
    <Box
      sx={{
        width: 170,
        display: 'flex',
        justifyContent: 'center',
        alignItems: 'center',
        svg: {
          width: '100%',
          height: '40px',
        },
      }}
      id={`color-box${id}`}
    ></Box>
  );
};

export default ColorScale;
