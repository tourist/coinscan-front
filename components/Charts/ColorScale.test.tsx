import { render, screen } from '@testing-library/react';

import ColorScale from './ColorScale';

afterEach(() => {
  jest.restoreAllMocks();
});

test('renders with single value', async () => {
  const { container } = render(<ColorScale id="color-scale-test" data={30} />);
  screen.getByText('+30%');
  expect(container).toMatchSnapshot();
});

test('renders with array', async () => {
  const { container } = render(
    <ColorScale id="color-scale-test" data={[-30, 70]} />
  );
  await screen.findByText('-30%');
  await screen.findByText('+70%');
  expect(container).toMatchSnapshot();
});

test('not render on 0 passed and show warning in console', async () => {
  const mockedWarn = jest.spyOn(console, 'warn').mockImplementation(() => {});
  const { container } = render(<ColorScale id="color-scale-test" data={0} />);
  expect(container).toMatchSnapshot();
  expect(mockedWarn).toBeCalledWith(
    "0 passed as data value, should be handled outside as special case as it's not presentable."
  );
});

test('issue error on array length passed', async () => {
  const mockedError = jest.spyOn(console, 'error').mockImplementation(() => {});
  const { container } = render(
    <ColorScale id="color-scale-test" data={[10, 20, 30]} />
  );
  expect(container).toMatchSnapshot();
  expect(mockedError).toBeCalledWith(
    'Wrong data array length. Expected 2 elements. Passed: 3'
  );
});
