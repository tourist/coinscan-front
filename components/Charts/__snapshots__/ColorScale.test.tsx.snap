// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`issue error on array length passed 1`] = `
<div>
  <div
    class="MuiBox-root css-11q2k9a"
    id="color-boxcolor-scale-test"
  />
</div>
`;

exports[`not render on 0 passed and show warning in console 1`] = `
<div>
  <div
    class="MuiBox-root css-11q2k9a"
    id="color-boxcolor-scale-test"
  />
</div>
`;

exports[`renders with array 1`] = `
<div>
  <div
    class="MuiBox-root css-11q2k9a"
    id="color-boxcolor-scale-test"
  >
    <svg
      background="black"
      preserveAspectRatio="xMidYMid"
      viewBox="0 0 150 20"
    >
      <defs>
        <clippath
          id="negative-clip-color-boxcolor-scale-test"
        >
          <rect
            height="100%"
            width="15%"
            x="35%"
            y="0"
          />
        </clippath>
        <lineargradient
          id="negative-gradient-color-boxcolor-scale-test"
          x1="0%"
          x2="100%"
          y1="0%"
          y2="0%"
        >
          <stop
            class="start"
            offset="0%"
            stop-color="red"
            stop-opacity="1"
          />
          <stop
            class="end"
            offset="100%"
            stop-color="#ff7377"
            stop-opacity="1"
          />
        </lineargradient>
        <clippath
          id="positive-clip-color-boxcolor-scale-test"
        >
          <rect
            height="100%"
            width="35%"
            x="50%"
            y="0"
          />
        </clippath>
        <lineargradient
          id="positive-gradient-color-boxcolor-scale-test"
          x1="0%"
          x2="100%"
          y1="0%"
          y2="0%"
        >
          <stop
            class="start"
            offset="0%"
            stop-color="#77dd77"
            stop-opacity="1"
          />
          <stop
            class="end"
            offset="100%"
            stop-color="#09b051"
            stop-opacity="1"
          />
        </lineargradient>
      </defs>
      <rect
        clip-path="url(#negative-clip-color-boxcolor-scale-test)"
        fill="url(#negative-gradient-color-boxcolor-scale-test)"
        height="100%"
        width="50%"
        x="0"
        y="0"
      />
      <text
        dominant-baseline="central"
        fill="#ffffff"
        font-size="0.7rem"
        font-weight="700"
        text-anchor="end"
        x="33%"
        y="50%"
      >
        -30%
      </text>
      <rect
        clip-path="url(#positive-clip-color-boxcolor-scale-test)"
        fill="url(#positive-gradient-color-boxcolor-scale-test)"
        height="100%"
        width="50%"
        x="50%"
        y="0"
      />
      <text
        dominant-baseline="central"
        fill="#ffffff"
        font-size="0.7rem"
        font-weight="700"
        text-anchor="end"
        x="83%"
        y="50%"
      >
        +70%
      </text>
    </svg>
  </div>
</div>
`;

exports[`renders with single value 1`] = `
<div>
  <div
    class="MuiBox-root css-11q2k9a"
    id="color-boxcolor-scale-test"
  >
    <svg
      background="black"
      preserveAspectRatio="xMidYMid"
      viewBox="0 0 150 20"
    >
      <defs>
        <clippath
          id="positive-clip-color-boxcolor-scale-test"
        >
          <rect
            height="100%"
            width="15%"
            x="50%"
            y="0"
          />
        </clippath>
        <lineargradient
          id="positive-gradient-color-boxcolor-scale-test"
          x1="0%"
          x2="100%"
          y1="0%"
          y2="0%"
        >
          <stop
            class="start"
            offset="0%"
            stop-color="#77dd77"
            stop-opacity="1"
          />
          <stop
            class="end"
            offset="100%"
            stop-color="#09b051"
            stop-opacity="1"
          />
        </lineargradient>
      </defs>
      <rect
        clip-path="url(#positive-clip-color-boxcolor-scale-test)"
        fill="url(#positive-gradient-color-boxcolor-scale-test)"
        height="100%"
        width="50%"
        x="50%"
        y="0"
      />
      <text
        dominant-baseline="central"
        fill="#ffffff"
        font-size="0.7rem"
        font-weight="700"
        text-anchor="start"
        x="67%"
        y="50%"
      >
        +30%
      </text>
    </svg>
  </div>
</div>
`;
